src/builder.rs

```diff
@@ -92,12 +91,20 @@ impl StatsdBuilder {
         interval: Duration,
     ) -> Result<Self, BuildError>
     where
-        T: AsRef<str>,
+        T: ToSocketAddrs,
     {
+        let endpoint = endpoint
+            .to_socket_addrs()
```

Comment:
I'm relatively new to Rust, but I believe this is where DNS resolution occurs, and after this we're just passing around a regular IPv4/v6 address. If the dogstatsd agent is running in a separate pod (k8s in my case) that gets re-deployed and the IP changes, we will stop sending metrics to the right place. 

Ideally, I think we'd propagate the `T: ToSocketAddrs` all the way down to the [`send_all`](https://github.com/remind101/metrics-exporter-dogstatsd/blob/2958118c3782817956ab77c28180993cf2d92db5/src/builder.rs#L432) method, which would require making `ExporterConfig` generic over `T: ToSocketAddrs` as well (or `Box<dyn ToSocketAddrs>`?). [`UdpSocket.send_to`](https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#method.send_to) would then handle the `.to_socket_addrs()` for us, re-resolving DNS as needed.

Reply:
Yes, that is what is doing the resolution.  I am also relatively new to rust.  I just needed to solve the problem of having it resolve, but your comment above sounds right.  However, I am not sure you want to necessarily re-resolve on each send as that gets expensive as well.   What *might* be better (and is a *lot* more work) would be to have another thread running which woke up every <interval> (probably based on the SOA lifetime of the DNS resolution?) and re-resolved that address.  For what I was accomplishing, either of those was overkill.    However, to your point, the current implementation which requires passing in the IP address at the top and doesn't resolve at all would still result in your sending stats if your k8s pod got redeployed and the ip address changed, no?

Reply:
So my solution doesn't change that, but at least allows providing a dns name rather than a static ip for configuration, so is (imho) strictly better than the current solution even if not perfect (such as for the case you suggested)

Reply:
Totally agree that at least being able to pass a DNS name is an improvement to start with! Though for what it's worth, you can do the DNS resolution outside this crate and pass the resolved address into `.with_push_gateway` without needing to use this fork:

```rust
let statsd_endpoint = host  // Host is something like "datadog-agent:8125"
    .to_socket_addrs()
    .expect("Invalid statsd push gateway endpoint")
    .next()
    .expect("to_socket_addrs returned an empty iterator")
    .to_string();

let statsd_builder = StatsdBuilder::new()
    .with_push_gateway(statsd_endpoint, Duration::from_secs(1))
    .expect("Could not set up statsd push gateway");

statsd_builder
    .install()
    .expect("Failed to install statsd recorder/exporter");
```

I think passing in an IP today that doesn't resolve to the right pod and giving a DNS name that initially resolves but eventually gets updated would have the same outcome, that `send_to` would error [here](https://github.com/dialtone/metrics-exporter-dogstatsd/blob/main/src/builder.rs#L447) (With one of [these](https://doc.rust-lang.org/nightly/std/io/enum.ErrorKind.html) variants, probably `HostUnreachable` if I had to guess).

You raise a good point about doing the DNS resolution too frequently. Digging through [hot-shots](https://github.com/brightcove/hot-shots) (a node dogstatsd library we also use), they let you set a `cacheDns` flag and a TTL up front, then on every send they check if the TTL has passed and re-resolve if so ([source](https://github.com/brightcove/hot-shots/blob/master/lib/transport.js#L56)). Doing that in the [loop in `build`](https://github.com/dialtone/metrics-exporter-dogstatsd/blob/main/src/builder.rs#L320) is probably a fair bit easier than a separate thread just to re-resolve the address.

Reply:
> Though for what it's worth, you can do the DNS resolution outside this crate and pass the resolved address into `.with_push_gateway` without needing to use this fork:

For sure.. but it seemed silly to me to resolve, remake it into a string and then re-parse it again.   Also, doing it this way allowed for IPV6 resolved addresses to work whereas I am not 100% certain if that would work in the above.  The original iteration of this code also only bound to the IPV4 address (0.0.0.0:0).. I changed that to bind to the IPV6 variant, which (at least on most OS's) also binds to the IPV4 version as far as I could tell.  I do think the re-resolve inside the loop on a dns cache timeout would be the overall best solution, and there are a number of systems which do that I believe.